cmd/save_test.go

```diff
@@ -199,6 +204,129 @@ func TestSaveRun(t *testing.T) {
+		},
+		{
+			"dataset file me ref",
+			"qri save --file dataset.yaml me/my_dataset",
+			"dataset saved: test_peer/my_dataset@/ipfs/QmSSTAcuehjFFiHNBFicTXJv2uhbKg4hjGgCL8W46D3jGG\nthis dataset has 1 validation errors\n",
+		},
+		{
+			"dataset file explicit ref",
+			"qri save --file dataset.yaml test_peer/my_dataset",
+			"dataset saved: test_peer/my_dataset@/ipfs/QmeW1M8W9yj69pxE3uJZbYVBCaEARYkrkje3NHY1s6tHtP\nthis dataset has 1 validation errors\n",
+		},
+		{
+			"body file infer name",
+			"qri save --body body_ten.csv",
+			"dataset saved: test_peer/body_tencsv@/ipfs/QmWaS8ndeAhAWWSUmwJ6HiXKMNWnsL1LYhzrWJyzCgR9pd\nthis dataset has 1 validation errors\n",
+		},
+		{
+			"body file me ref",
+			"qri save --body body_ten.csv me/my_dataset",
+			"dataset saved: test_peer/my_dataset@/ipfs/QmeyznW4FBaLY8rk5xpZyFHXixrq8qFJrFf8v1REZzmLLg\nthis dataset has 1 validation errors\n",
+		},
+		{
+			"body file explicit ref",
+			"qri save --body body_ten.csv test_peer/my_dataset",
+			"dataset saved: test_peer/my_dataset@/ipfs/QmP9yay2KwPnyDLLhGXZsLdKG31qMAR3CSBAmscz3znC9C\nthis dataset has 1 validation errors\n",
+		},
+		// TODO(dustmop): It's intended that a user can save a dataset with a structure but no
+		// body. At some point that functionality broke, because there was no test for it. Fix that
+		// in a follow-up change.
+		//{
+		//	"structure file me ref",
+		//	"qri save --file structure.json me/my_dataset",
+		//	"TODO(dustmop): Should be possible to save a dataset with structure and no body",
+		//},
+		//{
+		//	"structure file explicit ref",
+		//	"qri save --file structure.json test_peer/my_dataset",
+		//	"TODO(dustmop): Should be possible to save a dataset with structure and no body",
+		//},
+	}
+	for _, c := range goodCases {
+		// TODO(dustmop): Would be preferable to instead have a way to clear the refstore
+		run := NewTestRunner(t, "test_peer", "qri_test_save_basic")
+		defer run.Delete()
+
+		err := run.ExecCommand(c.command)
+		if err != nil {
+			t.Errorf("%s: error %s\n", c.description, err)
+			continue
+		}
+		actual := parseDatasetRefFromOutput(run.GetCommandOutput())
+		if diff := cmp.Diff(c.expect, actual); diff != "" {
+			t.Errorf("%s: result mismatch (-want +got):%s\n", c.description, diff)
+		}
+	}
+
+	badCases := []struct {
+		description string
+		command     string
+		expectErr   string
+	}{
+		{
+			"dataset file other username",
+			"qri save --file dataset.yaml other/my_dataset",
+			"cannot save using a different username than \"test_peer\"",
+		},
+		{
+			"dataset file explicit version",
+			"qri save --file dataset.yaml me/my_dataset@/ipfs/QmVersion",
+			"ref can only have username/name",
+		},
+		{
+			"body file other username",
+			"qri save --body body_ten.csv other/my_dataset",
+			"cannot save using a different username than \"test_peer\"",
+		},
+	}
+	for _, c := range badCases {
+		run := NewTestRunner(t, "test_peer", "qri_test_save_basic")
+		defer run.Delete()
```

Comment:
It's funny to think about what this defer statement is doing. Since [defers call on enclosing function return](https://golang.org/ref/spec#Defer_statements) this is stacking up a bunch of delete calls in a row then calling them all in succession after the test function returns.

No need to change it, but running this in a subtest _would_ alter this

Reply:
Whoops, the intention was actually to have that latter behavior, it was buggy before. Putting it into subtests fixes that.